File: java/src/FloydWarshall.java
public class FloydWarshall {

	static final int INF = Integer.MAX_VALUE / 2;

	// precondition: d[i][i] == 0
	public static int[][] floydWarshall(int[][] d) {
		int n = d.length;
		int[][] pred = new int[n][n];
		for (int i = 0; i < n; i++)
			for (int j = 0; j < n; j++)
				pred[i][j] = (i != j && d[i][j] != INF) ? i : -1;
		for (int k = 0; k < n; k++) {
			for (int i = 0; i < n; i++) {
				if (d[i][k] == INF)
					continue;
				for (int j = 0; j < n; j++) {
					if (d[k][j] == INF)
						continue;
					if (d[i][j] > d[i][k] + d[k][j]) {
						d[i][j] = d[i][k] + d[k][j];
						d[i][j] = Math.max(d[i][j], -INF);
						pred[i][j] = pred[k][j];
					}
				}
			}
		}
		for (int i = 0; i < n; i++)
			if (d[i][i] < 0)
				return null;
		return pred;
	}

	// Usage example
	public static void main(String[] args) {
		int[][] dist = {{0, 3, 2}, {0, 0, 1}, {INF, 0, 0}};
		int[][] pred = floydWarshall(dist);

		System.out.println(0 == dist[0][0]);
		System.out.println(2 == dist[0][1]);
		System.out.println(2 == dist[0][2]);
		System.out.println(-1 == pred[0][0]);
		System.out.println(2 == pred[0][1]);
		System.out.println(0 == pred[0][2]);
	}
}

File: golang/graphs/floyd_warshall.go
package graphs

import "math"

const INF = math.MaxInt32 / 2

// FloydWarshall computes all-pairs shortest paths in O(V^3)
// Modifies dist in-place. Returns predecessor matrix, or nil if negative cycle exists.
// https://en.wikipedia.org/wiki/Floyd-Warshall_algorithm
func FloydWarshall(dist [][]int) [][]int {
	n := len(dist)
	pred := make([][]int, n)
	for i := range pred {
		pred[i] = make([]int, n)
		for j := range pred[i] {
			if i == j || dist[i][j] == INF {
				pred[i][j] = -1
			} else {
				pred[i][j] = i
			}
		}
	}

	for k := 0; k < n; k++ {
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				if dist[i][j] > dist[i][k]+dist[k][j] {
					dist[i][j] = dist[i][k] + dist[k][j]
					pred[i][j] = pred[k][j]
				}
			}
		}
	}

	for i := 0; i < n; i++ {
		if dist[i][i] < 0 {
			return nil
		}
	}
	return pred
}
